### PR TITLE
Shared Axis Transition

### PR DESCRIPTION
**Pull request by shihaohong**
## Description
Implement the shared x-axis and shared y-axis transitions. This PR also deletes the existing `SharedZAxisTransitionsBuilder` and pulls it into `SharedAxisPageTransitionsBuilder` through an option called `SharedAxisTransitionType.scaled`.

```dart
MaterialApp(
  theme: ThemeData(
    pageTransitionsTheme: PageTransitionsTheme(
      builders: {
        TargetPlatform.android: SharedAxisPageTransitionsBuilder(
          transitionType: SharedAxisTransitionType.horizontal, // or SharedAxisTransitionType.vertical
        ),
      },
    ),
  ),
  // ...
);
```

### Shared X-Axis Transition
Shared X-axis Transition

### Shared Y-Axis Transition
Shared Y-axis Transition

### Similarity to Shared Z Axis Transition
I noticed that the curves used for this transition is identical to https://github.com/flutter/packages/pull/57 minus the difference in scale/translation animations, but decided to create a separate builder and suite of tests in the event that the specs change, which might make it difficult to decouple their implementations. If using the same builder and adding a new enum value of something like `SharedAxisTransition.scaled` seems more convenient, we could implement it that way as well.

## Sample App:
```dart
import 'package:flutter/material.dart';
import 'package:animations/animations.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _selectedIndex = 0;

  final List<Color> _colors = [Colors.white, Colors.red, Colors.yellow];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Page Transition Example'),
      ),
      body: PageTransitionSwitcher(
        transitionBuilder: (
          Widget child,
          Animation<double> primaryAnimation,
          Animation<double> secondaryAnimation,
        ) {
          return SharedAxisTransition(
            animation: primaryAnimation,
            secondaryAnimation: secondaryAnimation,
            child: child,
            transitionType: SharedAxisTransitionType.vertical,
          );
        },
        child: Container(
          key: ValueKey<int>(_selectedIndex),
          color: _colors[_selectedIndex],
          child: Center(
            child: FlutterLogo(size: 300),
          )
        ),
      ),
      bottomNavigationBar: BottomNavigationBar(
        items: const <BottomNavigationBarItem>[
          BottomNavigationBarItem(
            icon: Icon(Icons.home),
            title: Text('White'),
          ),
          BottomNavigationBarItem(
            icon: Icon(Icons.business),
            title: Text('Red'),
          ),
          BottomNavigationBarItem(
            icon: Icon(Icons.school),
            title: Text('Yellow'),
          ),
        ],
        currentIndex: _selectedIndex,
        onTap: (int index) {
          setState(() {
            _selectedIndex = index;
          });
        },
      ),
    );
  }
}
```

### CODE REVIEW

File: packages/animations/lib/src/shared_axis_transition.dart [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+// found in the LICENSE file.
+
+import 'package:flutter/animation.dart';
+import 'package:flutter/foundation.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter/widgets.dart';
+
+import 'utils/curves.dart';
+
+/// Determines which type shared axis transition.
```
Comment (goderbauer) [inline packages/animations/lib/src/shared_axis_transition.dart]: nit: this sentence seems to be incomplete. Maybe missing an "of"?

File: packages/animations/lib/src/shared_axis_transition.dart [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+///
+/// Consider using [SharedAxisTransition] within a
+/// [PageTransitionsTheme] if you want to apply this kind of transition to
+/// [MaterialPageRoute] transitions within a Navigator (see
+/// [SharedAxisPageTransitionsBuilder] for example code).
+///
+/// This transition can also be used directly in a
+/// [PageTransitionSwitcher.transitionBuilder] to transition
+/// from one widget to another as seen in the following example:
+/// ```dart
```
Comment (goderbauer) [inline packages/animations/lib/src/shared_axis_transition.dart]: nit: add a blank line before this

File: packages/animations/lib/src/shared_axis_transition.dart [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+  /// The animation that transitions [child] when new content is pushed on top
+  /// of it.
+  ///
+  /// See also:
+  ///
+  ///  * [TransitionRoute.secondaryAnimation], which is the value given to this
+  ///    property when the it is used as a page transition.
+  final Animation<double> secondaryAnimation;
+
+  /// Determines which type shared axis transition.
```
Comment (goderbauer) [inline packages/animations/lib/src/shared_axis_transition.dart]: nit: incomplete sentence?

File: packages/animations/lib/src/shared_axis_transition.dart [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+  /// Determines which [SharedAxisTransitionType] to build.
+  final SharedAxisTransitionType transitionType;
+
+  @override
+  Widget buildTransitions<T>(
+      PageRoute<T> route,
+      BuildContext context,
+      Animation<double> animation,
+      Animation<double> secondaryAnimation,
+      Widget child) {
```
Comment (goderbauer) [inline packages/animations/lib/src/shared_axis_transition.dart]: nit: add a trailing comma and reformat.

File: packages/animations/lib/src/shared_axis_transition.dart [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+  final Widget child;
+
+  static Animatable<double> fadeInTransition =
+      CurveTween(curve: decelerateEasing)
+          .chain(CurveTween(curve: const Interval(0.3, 1.0)));
+
+  @override
+  Widget build(BuildContext context) {
+    final Animatable<Offset> slideInTransition = Tween<Offset>(
+      begin: transitionType == SharedAxisTransitionType.horizontal
```
Comment (goderbauer) [inline packages/animations/lib/src/shared_axis_transition.dart]: enums are best used with switch/case

File: packages/animations/lib/src/shared_axis_transition.dart [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+
+  static Animatable<double> fadeOutTransition =
+      FlippedCurveTween(curve: accelerateEasing)
+          .chain(CurveTween(curve: const Interval(0.0, 0.3)));
+
+  @override
+  Widget build(BuildContext context) {
+    final Animatable<Offset> slideOutTransition = Tween<Offset>(
+      begin: Offset.zero,
+      end: transitionType == SharedAxisTransitionType.horizontal
```
Comment (goderbauer) [inline packages/animations/lib/src/shared_axis_transition.dart]: same

File: packages/animations/test/shared_axis_transition_test.dart [hunk 7]
```diff
@@ -0,0 +1,1077 @@
...
+        );
+
+        expect(find.text(bottomRoute), findsOneWidget);
+        expect(
+            _getTranslationOffset(
+              bottomRoute,
+              tester,
+              SharedAxisTransitionType.horizontal,
+            ),
+            0.0);
```
Comment (goderbauer) [inline packages/animations/test/shared_axis_transition_test.dart]: uber-nit (feel free to ignore): I think that dartfmt formats these multi-line expects nicer if you add a trailing comma.

(here and elsewhere)

**Comment by shihaohong** [comment]
@goderbauer I addressed the code review feedback. 

We also spoke offline about having the SharedZAxisTransition builder simply be a part of the x and y SharedAxisTransition builders. I've deleted the SharedZAxis transition code and incorporated it into the SharedAxisTransition builder itself.

File: packages/animations/lib/src/shared_axis_transition.dart [hunk 8]
```diff
@@ -9,18 +9,27 @@ import 'package:flutter/widgets.dart';
 
 import 'utils/curves.dart';
 
-/// An in-place fade and scale transition used by [PageTransitionsTheme]
-/// to create a page transition with [SharedZAxisTransition].
+/// Determines which type of shared axis transition is used.
+enum SharedAxisTransitionType {
+  /// Creates a shared axis vertical translation page transition.
```
Comment (goderbauer) [inline packages/animations/lib/src/shared_axis_transition.dart]: Can you mention the alternative name (x, y, z axis) in these doc strings?

File: packages/animations/lib/src/shared_axis_transition.dart [hunk 9]
```diff
@@ -48,11 +61,11 @@ import 'utils/curves.dart';
 ///         ),
 ///       );
 ///     },
-///     '/a': (BuildContext context) {
+///     '/a' : (BuildContext context) {
 ///       return Container(
 ///         color: Colors.blue,
 ///         child: Center(
-///           child: RaisedButton(
+///           child: MaterialButton(
```
Comment (goderbauer) [inline packages/animations/lib/src/shared_axis_transition.dart]: I believe MaterialButton is to be avoided and RaisedButton, FlatButton, or OutlineButton should be used.

File: packages/animations/pubspec.yaml [hunk 10]
```diff
@@ -9,6 +9,7 @@ environment:
 dependencies:
   flutter:
     sdk: flutter
+  vector_math: ^2.0.8
```
Comment (goderbauer) [inline packages/animations/pubspec.yaml]: This is only used in tests, right? In that case, it should be a dev_dependency and moved to the section below.

**Pull request merged**